File: scala/2015/day04b/Program.scala
import scala.sys.exit

import java.security.MessageDigest

object Program {
  def usage(): Unit = {
    Console.err.println("usage: scala Program <input file>")
    sys.exit(1)
  }

  def process(key: String): Int = {
    val md5 = MessageDigest.getInstance("MD5")
    def checkKey(key: String, n: Int): Int = {
      val tryKey = key ++ n.toString
      val digestBytes = md5.digest(tryKey.getBytes("UTF-8"))
      val hexDigest = digestBytes.map("%02x".format(_)).mkString
      if ( hexDigest.startsWith("000000") ) n else checkKey(key, n + 1)
    }
    checkKey(key, 1)
  }

  def main(args: Array[String]): Unit = {
    args.toList match {
      case key :: Nil =>
        val result = process(key)
        println(s"result = $result")
      case _ =>
        usage()
    }
  }
}

File: elixir/2015/day04b/Program.exs
#!/usr/bin/env elixir

defmodule Program do
  def main(args) do
    case args do
      [key] ->
        result = process(key)
        IO.puts("result = #{result}")
      _ ->
        usage()
    end
  end

  defp usage do
    IO.puts(:stderr, "usage: Program.exs <input file>")
    System.halt(1)
  end

  defp process(key) do
     find_value(key, 1)
  end

  defp find_value(key, n) do
    try_key = key <> Integer.to_string(n)
    digest = :crypto.hash(:md5, try_key) |> Base.encode16(case: :lower)
    if String.starts_with?(digest, "000000") do
      n
    else
      find_value(key, n + 1)
    end
  end
end

Program.main(System.argv())
